consensus/src/units/store.rs

```diff
@@ -1,291 +1,273 @@
-use super::*;
-use itertools::Itertools;
-use log::{trace, warn};
-use std::{collections::HashSet, fmt};
-
-#[derive(Clone, Eq, PartialEq, Hash)]
-pub struct UnitStoreStatus<'a> {
-    index: NodeIndex,
-    forkers: &'a NodeSubset,
+use std::{
+    collections::HashMap,
+    fmt::{Display, Formatter, Result as FmtResult},
+};
+
+use crate::{
+    units::{HashFor, Unit, UnitCoord},
+    NodeCount, NodeIndex, NodeMap, Round,
+};
+
+/// An overview of what is in the unit store.
+pub struct UnitStoreStatus {
     size: usize,
-    height: Round,
     top_row: NodeMap<Round>,
-    first_missing_rounds: NodeMap<Round>,
 }
 
-impl<'a> UnitStoreStatus<'a> {
-    fn new(
-        index: NodeIndex,
-        forkers: &'a NodeSubset,
-        size: usize,
-        height: Round,
-        top_row: NodeMap<Round>,
-        first_missing_rounds: NodeMap<Round>,
-    ) -> Self {
-        Self {
-            index,
-            forkers,
-            size,
-            height,
-            top_row,
-            first_missing_rounds,
-        }
-    }
-
-    pub fn rounds_behind(&self) -> Round {
-        self.height
-            .saturating_sub(self.top_row.get(self.index).cloned().unwrap_or(0))
+impl UnitStoreStatus {
+    /// Highest round among units in the store.
+    pub fn top_round(&self) -> Round {
+        self.top_row.values().max().cloned().unwrap_or(0)
     }
 }
 
-impl<'a> fmt::Display for UnitStoreStatus<'a> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "DAG size - {}; DAG height - {}", self.size, self.height)?;
-        if self.first_missing_rounds.item_count() > 0 {
-            write!(
-                f,
-                "; DAG first missing rounds - {}",
-                self.first_missing_rounds
-            )?;
-        }
-        write!(f, "; DAG top row - {}", self.top_row)?;
-        if !self.forkers.is_empty() {
-            write!(f, "; forkers - {}", self.forkers)?;
-        }
-        Ok(())
+impl Display for UnitStoreStatus {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "total units: {}, top row: {}", self.size, self.top_row)
     }
 }
 
-/// A component for temporarily storing units before they are declared "legit" and sent
-/// to the Terminal. We refer to the documentation https://cardinal-cryptography.github.io/AlephBFT/internals.html
-/// Section 5.4 for a discussion of this component and the notion of "legit" units.
-
-pub(crate) struct UnitStore<H: Hasher, D: Data, K: Keychain> {
-    by_coord: HashMap<UnitCoord, SignedUnit<H, D, K>>,
-    by_hash: HashMap<H::Hash, SignedUnit<H, D, K>>,
-    parents: HashMap<H::Hash, Vec<H::Hash>>,
-    //the number of unique nodes that we hold units for a given round
-    is_forker: NodeSubset,
-    legit_buffer: Vec<SignedUnit<H, D, K>>,
-    max_round: Round,
+/// Stores units and remembers the first instance of any unit with a specific coordinate inserted.
```

Comment:
I would also include in this comment info that first instance == canonical, and to the first part, that it stores units by hash. Then everything will be 100% clear

Reply:
Yup, added.

Reply:
I meant changing the doc-string so that it contains the definition of `canonical unit`, and that it contains info that all the units are stored by hash.
It's a very minor thing tho, feel free to ignore & resolve.

Reply:
Better now?